kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -233,48 +288,148 @@ class KinesisSink private (
     ()
   }
 
-  // TODO: limit max retries?
-  def sendBatch(batch: List[(ByteBuffer, String)], nextBackoff: Long = minBackoff): Unit =
-    if (batch.size > 0) {
+  /**
+   *  Max number of retries is unlimitted, so when Kinesis stream is under heavy load,
+   *   the events accumulate in collector memory for later retries. The fix for this is to use
+   *   sqs queue as a buffer and sqs2kinesis to move events back from sqs queue to kinesis stream.
+   *  Consider using sqs buffer in heavy load scenarios.
+   *
+   */
+  def sendBatch(batch: List[BatchEntry], nextBackoff: Long = minBackoff): Unit =
+    if (batch.nonEmpty) {
       log.info(s"Writing ${batch.size} Thrift records to Kinesis stream ${streamName}")
-      val putData = for {
-        p <- multiPut(streamName, batch)
-      } yield p
 
-      putData onComplete {
+      multiPut(streamName, batch).onComplete {
         case Success(s) => {
           val results = s.getRecords.asScala.toList
           val failurePairs = batch zip results filter { _._2.getErrorMessage != null }
           log.info(
             s"Successfully wrote ${batch.size - failurePairs.size} out of ${batch.size} records"
           )
-          if (failurePairs.size > 0) {
+          if (failurePairs.nonEmpty) {
             failurePairs.foreach(
               f =>
                 log.error(
                   s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]"
                 )
             )
-            log.error(s"Retrying all failed records in $nextBackoff milliseconds...")
             val failures = failurePairs.map(_._1)
-            scheduleBatch(failures, nextBackoff)
+            val retryErrorMsg = s"Retrying all failed records in $nextBackoff milliseconds..."
+            sendToSqsOrRetryToKinesis(failures, nextBackoff)(retryErrorMsg)
           }
         }
         case Failure(f) => {
           log.error("Writing failed.", f)
-          log.error(s"Retrying in $nextBackoff milliseconds...")
-          scheduleBatch(batch, nextBackoff)
+          val retryErrorMsg = s"Retrying in $nextBackoff milliseconds..."
+          sendToSqsOrRetryToKinesis(batch, nextBackoff)(retryErrorMsg)
         }
       }
     }
 
-  private def multiPut(name: String, batch: List[(ByteBuffer, String)]): Future[PutRecordsResult] =
+  private def sendToSqsOrRetryToKinesis(
+    failures: List[BatchEntry],
+    nextBackoff: Long
+  )(
+    retryErrorMsg: String
+  ): Unit =
+    maybeSqs match {
+      case Some(sqs) =>
+        log.info(
+          s"Sending ${failures.size} events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+        )
+        putToSqs(sqs, failures)
```

Comment:
Don't we want to add a callback with a logging of a potential error, just in case?

Reply:
It's done in retry mechanism in `sendToSqsWithRetry`, so further in function calls.